modules/module_service.go

```diff
@@ -86,11 +86,10 @@ func (w *moduleService) start(serviceContext context.Context) error {
 	}
 
 	err = w.service.AwaitRunning(serviceContext)
-	if errors.Is(serviceContext.Err(), context.Canceled) && errors.Is(err, context.Canceled) {
-		// We were asked to cancel, but the underlying service may continue starting and complete its startup after we return from moduleService.start.
-		// If we return an error, then we would be considered Failed and won't have moduleService.stop invoked.
-		// We need to invoke moduleService.stop to stop the underlying service in case it finishes starting successfully later.
-		return nil
+	if err != nil {
+		// Make sure that underlying service is stopped before returning
+		// (eg. in case of context cancellation, AwaitRunning returns early, but service may still be starting).
+		_ = services.StopAndAwaitTerminated(context.Background(), w.service)
 	}
 	return err
```

Comment:
```suggestion
	return errors.Wrapf(err, "waiting for service %s to start", w.name)
```

Reply:
👍 Makes sense. I was thinking the same when writing the test.